File: packages/ws-proto-transport/src/interceptor.ts
import type { Message } from "@bufbuild/protobuf";
import type { GenMessage } from "@bufbuild/protobuf/codegenv2";

/**
 * Interceptors are the typed, message-level extension point of the client
 * (the analog of connect-web interceptors). They wrap a typed RPC call and can
 * read/modify request metadata, observe or transform the request and response
 * MESSAGES, short-circuit a call, and read response trailers — across all four
 * method kinds. They are installed transport-wide via WsTransportOptions.
 *
 * This module defines only the types and the composer; WsTransport.unary/stream
 * build the chain and supply the terminal function that does the actual I/O.
 */

/** MethodKind mirrors the protobuf method kinds. */
export type MethodKind =
  | "unary"
  | "server_streaming"
  | "client_streaming"
  | "bidi_streaming";

/**
 * MethodInfo describes one RPC method to the transport and interceptors. The
 * generated client supplies it (service type name, method name, kind, and the
 * input/output protobuf schemas used for (de)serialization).
 */
export interface MethodInfo<I extends Message = Message, O extends Message = Message> {
  readonly typeName: string; // fully-qualified service name, e.g. "echo.v1.EchoService"
  readonly name: string; // method name, e.g. "Unary"
  readonly kind: MethodKind;
  readonly input: GenMessage<I>;
  readonly output: GenMessage<O>;
}

/** route returns the wire path "/pkg.Service/Method" for a method. */
export function route(method: MethodInfo): string {
  return `/${method.typeName}/${method.name}`;
}

/** A unary RPC invocation as seen by interceptors (one request message). */
export interface UnaryRequest<I extends Message = Message, O extends Message = Message> {
  readonly stream: false;
  readonly method: MethodInfo<I, O>;
  /** Request metadata; interceptors may mutate it before calling next. */
  header: Record<string, string>;
  readonly message: I;
  readonly signal?: AbortSignal;
  readonly timeoutMs?: number;
}

/** A streaming RPC invocation as seen by interceptors (a stream of requests). */
export interface StreamRequest<I extends Message = Message, O extends Message = Message> {
  readonly stream: true;
  readonly method: MethodInfo<I, O>;
  header: Record<string, string>;
  readonly message: AsyncIterable<I>;
  readonly signal?: AbortSignal;
  readonly timeoutMs?: number;
}

/** The completed unary response (leading header, single message, trailer). */
export interface UnaryResponse<I extends Message = Message, O extends Message = Message> {
  readonly stream: false;
  readonly method: MethodInfo<I, O>;
  header: Record<string, string>;
  message: O;
  trailer: Record<string, string>;
}

/** The streaming response (leading header, a stream of messages, trailer). */
export interface StreamResponse<I extends Message = Message, O extends Message = Message> {
  readonly stream: true;
  readonly method: MethodInfo<I, O>;
  header: Record<string, string>;
  message: AsyncIterable<O>;
  trailer: Record<string, string>;
}

export type AnyRequest = UnaryRequest | StreamRequest;
export type AnyResponse = UnaryResponse | StreamResponse;

/** AnyFn performs one RPC call; interceptors wrap it. */
export type AnyFn = (req: AnyRequest) => Promise<AnyResponse>;

/** Interceptor wraps an AnyFn to add behavior around a call. */
export type Interceptor = (next: AnyFn) => AnyFn;

/**
 * applyInterceptors composes interceptors around a terminal AnyFn. The FIRST
 * interceptor in the array is the OUTERMOST (runs first on the way in, last on
 * the way out) — matching connect-web ordering. With none, the terminal is
 * returned unchanged.
 */
export function applyInterceptors(next: AnyFn, interceptors?: readonly Interceptor[]): AnyFn {
  if (!interceptors || interceptors.length === 0) {
    return next;
  }
  return interceptors.reduceRight<AnyFn>((n, i) => i(n), next);
}

File: wsrpc/interceptor.go
package wsrpc

import (
	"context"

	"google.golang.org/protobuf/proto"
)

// StreamKind classifies a method so interceptors can branch on call shape.
type StreamKind int

const (
	StreamKindUnary StreamKind = iota
	StreamKindServerStream
	StreamKindClientStream
	StreamKindBidiStream
)

// MethodSpec describes one RPC method to the dispatch layer and interceptors.
type MethodSpec struct {
	Route string // "/pkg.Service/Method"
	Kind  StreamKind
}

// AnyRequest is the unary request as seen by interceptors. Header is mutable
// (add auth/trace metadata before the call); Msg is the request message — a
// pointer, so its fields may be mutated, or it may be replaced entirely.
type AnyRequest struct {
	Spec   MethodSpec
	Header map[string]string
	Msg    proto.Message
}

// AnyResponse is the completed unary response. Interceptors may read/replace Msg
// and read the leading Header / Trailer.
type AnyResponse struct {
	Spec    MethodSpec
	Header  map[string]string // leading response metadata (KIND_HEADER)
	Trailer map[string]string // END trailers
	Msg     proto.Message
}

// UnaryFunc performs one unary RPC; interceptors wrap it.
type UnaryFunc func(ctx context.Context, req *AnyRequest) (*AnyResponse, error)

// StreamingClientConn drives one streaming RPC. The terminal implementation
// wraps a *Stream; interceptors may wrap it to observe/modify Send/Receive.
type StreamingClientConn interface {
	Spec() MethodSpec
	RequestHeader() map[string]string
	Send(proto.Message) error
	CloseRequest() error
	Receive(proto.Message) error
	ResponseHeader() map[string]string
	ResponseTrailer() map[string]string
	CloseResponse() error
}

// StreamingClientFunc opens a streaming RPC; interceptors wrap it.
type StreamingClientFunc func(ctx context.Context, spec MethodSpec, header map[string]string) (StreamingClientConn, error)

// Interceptor wraps unary and streaming client calls. Implement both methods, or
// embed one of the adapter funcs below to wrap only one call shape.
type Interceptor interface {
	WrapUnary(UnaryFunc) UnaryFunc
	WrapStreamingClient(StreamingClientFunc) StreamingClientFunc
}

// UnaryInterceptorFunc adapts a plain unary wrapper into an Interceptor that
// leaves streaming calls untouched.
type UnaryInterceptorFunc func(UnaryFunc) UnaryFunc

func (f UnaryInterceptorFunc) WrapUnary(next UnaryFunc) UnaryFunc { return f(next) }
func (f UnaryInterceptorFunc) WrapStreamingClient(next StreamingClientFunc) StreamingClientFunc {
	return next
}

// StreamInterceptorFunc adapts a plain streaming wrapper into an Interceptor
// that leaves unary calls untouched.
type StreamInterceptorFunc func(StreamingClientFunc) StreamingClientFunc

func (f StreamInterceptorFunc) WrapUnary(next UnaryFunc) UnaryFunc { return next }
func (f StreamInterceptorFunc) WrapStreamingClient(next StreamingClientFunc) StreamingClientFunc {
	return f(next)
}

// chainUnary composes interceptors around a terminal UnaryFunc. The FIRST
// interceptor is the OUTERMOST (runs first in, last out).
func chainUnary(interceptors []Interceptor, terminal UnaryFunc) UnaryFunc {
	for i := len(interceptors) - 1; i >= 0; i-- {
		terminal = interceptors[i].WrapUnary(terminal)
	}
	return terminal
}

// chainStreaming composes interceptors around a terminal StreamingClientFunc,
// outermost-first.
func chainStreaming(interceptors []Interceptor, terminal StreamingClientFunc) StreamingClientFunc {
	for i := len(interceptors) - 1; i >= 0; i-- {
		terminal = interceptors[i].WrapStreamingClient(terminal)
	}
	return terminal
}
